quic/s2n-quic/src/provider/event/console_perf.rs

```diff
@@ -0,0 +1,449 @@
+    /// and durations are output as microseconds
+    TSV_RAW,
+}
+
+impl Format {
+    fn print_suffix(&self) -> bool {
+        match self {
+            Self::TSV => true,
+            Self::TSV_RAW => false,
+        }
+    }
+}
+
+pub struct Builder {
+    frequency: Option<Duration>,
+    format: Format,
+}
+
+impl Default for Builder {
+    fn default() -> Self {
+        Self {
+            frequency: None,
+            format: Format::TSV,
+        }
+    }
+}
+
+impl Builder {
+    /// Sets the frequency at which performance metrics will be printed
+    ///
+    /// This uses `tokio::spawn` to spawn metric printing task,
+    /// and thus must be called within a Tokio runtime
+    ///
+    /// If frequency is not set, the caller is responsible for calling
+    /// `print_header` and `print` at the required frequency.
+    pub fn with_frequency(mut self, frequency: Duration) -> Self {
+        self.frequency = Some(frequency);
+        self
+    }
+
+    /// Sets the format that performance metrics will be printed in
+    pub fn with_format(mut self, format: Format) -> Self {
+        self.format = format;
+        self
+    }
+
+    /// Builds the [`console_perf::Subscriber`]
+    pub fn build(self) -> Subscriber {
+        Subscriber {
+            counters: Default::default(),
+            frequency: self.frequency,
+            format: self.format,
+            spawned: self.frequency.is_none(),
+        }
+    }
+}
+
+impl Subscriber {
+    /// Prints the current performance metrics to the console via STDOUT
+    pub fn print(&self) {
+        self.counters.print(self.format);
+    }
+
+    /// Prints the names of each performance metric to the console via STDOUT
+    pub fn print_header(&self) {
+        self.counters.print_header(self.format);
+    }
+
+    fn spawn(&mut self) {
+        if let (false, Some(frequency)) = (self.spawned, self.frequency) {
+            self.spawned = true;
+            self.print_header();
+            let subscriber = self.clone();
+            tokio::spawn(async move {
+                loop {
+                    tokio::time::sleep(frequency).await;
+                    subscriber.print();
+                }
+            });
+        }
+    }
+}
+
+impl event::Subscriber for Subscriber {
+    type ConnectionContext = ();
+
+    #[inline]
+    fn create_connection_context(
+        &mut self,
+        _meta: &event::ConnectionMeta,
+        _info: &event::ConnectionInfo,
+    ) -> Self::ConnectionContext {
+        // Initialize the counters.last_updated_micros with the current
+        // time if it has not already been initialized
+        let _ = self.counters.last_updated_micros.compare_exchange(
+            0,
+            SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .expect("Clock may have gone backwards")
+                .as_micros() as u64,
+            Ordering::Relaxed,
+            Ordering::Relaxed,
+        );
```

Comment:
This should really be done in the spawn function, right? Doing a compare_exchange plus querying the current time per connection doesn't seem needed.

Reply:
I could move it to `spawn()`, though that is called for each connection context anyway. I wanted a place to initialize the `last_updated_micros` that would give a useful result for the first time that `print` is called after a connection starts. I'm thinking of the case where a server starts and no connections happen for a while, if I initialize `last_updated_micros` from server start, then the first metrics will have very low tx/rx rates

Reply:
Well if you put it behind the `spawned` check then it would only be called once

Reply:
ah, but that would only help with the auto-spawn case, which I'm removing.

Reply:
well it could still have a simple `init: bool` to track if the clock has been initialized, right?